In [4]:
pip install rfit


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.9/11.9 MB 21.1 MB/s eta 0:00:00a 0:00:01
  Preparing metadata (setup.py) ... done
  Created wheel for mysql-connector: filename=mysql_connector-2.2.9-cp39-cp39-macosx_10_9_x86_64.whl size=247957 sha256=1cecb6ef7d5db67f98b183193e0f4ea8f6293ef5ec0b6bd595f1382fe622c5d8
  Stored in directory: /Users/manishatherupalli/Library/Caches/pip/wheels/7b/14/39/5aad423666e827dfe9a1fbcd111ac17171e7c9865d570780ce
Successfully built mysql-connector
Note: you may need to restart the kernel to use updated packages.


Initital Dataset Cleaning and Manipulation 

In [2]:
import pandas as pd
import numpy as np
import rfit
import os

In [3]:
dataset = pd.read_csv('/Users/manishatherupalli/Desktop/ML FINAL PROJECT/Dataset/OnlineNewsPopularity.csv')

In [4]:
# Get summary statistics for each column
summary = dataset.describe()

# Get the correlation matrix for the data
corr_matrix = dataset.corr()

# Compute the average number of shares for each article category
avg_shares_by_category = dataset.groupby('data_channel_is_lifestyle')['shares'].mean()

# Compute the average number of shares for each day of the week
dataset['day_of_week'] = pd.to_datetime(dataset['timedelta'], unit='s').dt.dayofweek
avg_shares_by_day = dataset.groupby('day_of_week')['shares'].mean()

# Print out the results
print(summary)
print(corr_matrix)
print(avg_shares_by_category)
print(avg_shares_by_day)

          timedelta  n_tokens_title  n_tokens_content  n_unique_tokens  \
count  39644.000000    39644.000000      39644.000000     39644.000000   
mean     354.530471       10.398749        546.514731         0.548216   
std      214.163767        2.114037        471.107508         3.520708   
min        8.000000        2.000000          0.000000         0.000000   
25%      164.000000        9.000000        246.000000         0.470870   
50%      339.000000       10.000000        409.000000         0.539226   
75%      542.000000       12.000000        716.000000         0.608696   
max      731.000000       23.000000       8474.000000       701.000000   

       n_non_stop_words  n_non_stop_unique_tokens     num_hrefs  \
count      39644.000000              39644.000000  39644.000000   
mean           0.996469                  0.689175     10.883690   
std            5.231231                  3.264816     11.332017   
min            0.000000                  0.000000      0.000000  

In [5]:
rfit.dfchk(dataset)


dataframe Basic Check function -

1: info(): 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39644 entries, 0 to 39643
Data columns (total 63 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   url                            39644 non-null  object 
 1   timedelta                      39644 non-null  float64
 2   n_tokens_title                 39644 non-null  float64
 3   n_tokens_content               39644 non-null  float64
 4   n_unique_tokens                39644 non-null  float64
 5   n_non_stop_words               39644 non-null  float64
 6   n_non_stop_unique_tokens       39644 non-null  float64
 7   num_hrefs                      39644 non-null  float64
 8   num_self_hrefs                 39644 non-null  float64
 9   num_imgs                       39644 non-null  float64
 10  num_videos                     39644 non-null  float64
 11  average_token_length           39644 non-null  float64
 12 

In [6]:
dataset.columns

Index(['url', 'timedelta', 'n_tokens_title', 'n_tokens_content',
       'n_unique_tokens', 'n_non_stop_words', 'n_non_stop_unique_tokens',
       'num_hrefs', 'num_self_hrefs', 'num_imgs', 'num_videos',
       'average_token_length', 'num_keywords', 'data_channel_is_lifestyle',
       'data_channel_is_entertainment', 'data_channel_is_bus',
       'data_channel_is_socmed', 'data_channel_is_tech',
       'data_channel_is_world', 'kw_min_min', 'kw_max_min', 'kw_avg_min',
       'kw_min_max', 'kw_max_max', 'kw_avg_max', 'kw_min_avg', 'kw_max_avg',
       'kw_avg_avg', 'self_reference_min_shares', 'self_reference_max_shares',
       'self_reference_avg_sharess', 'weekday_is_monday', 'weekday_is_tuesday',
       'weekday_is_wednesday', 'weekday_is_thursday', 'weekday_is_friday',
       'weekday_is_saturday', 'weekday_is_sunday', 'is_weekend', 'LDA_00',
       'LDA_01', 'LDA_02', 'LDA_03', 'LDA_04', 'global_subjectivity',
       'global_sentiment_polarity', 'global_rate_positive_words',
     

Total 61 attributes and 39644 rows
No Missing values and No Nulls

In [7]:
# First identifying what columns have these extra spaces
bad_columns = [x for x in dataset.columns if x.endswith(' ') or x.startswith(' ')]
print('Number of Columns with unwanted spaces: ',len(bad_columns))



Number of Columns with unwanted spaces:  0


In [8]:
dataset.columns = dataset.columns.str.strip()

# Checking to see if the problem is resolved
bad_columns_validation = [x for x in dataset.columns if x.endswith(' ') or x.startswith(' ')]
print('\nAfter Fix:\nNumber of Columns still with issue: ',len(bad_columns_validation))


After Fix:
Number of Columns still with issue:  0


In [9]:
# Running the code again

dataset_viz = dataset.copy()

dataset_viz['Data_Channel'] = np.where(dataset_viz['data_channel_is_lifestyle']==1,'Lifestyle',np.where(dataset_viz['data_channel_is_entertainment']==1,"Entertainment",np.where(dataset_viz['data_channel_is_bus']==1,'Business',np.where(dataset_viz['data_channel_is_socmed']==1,'Social Media',np.where(dataset_viz['data_channel_is_tech']==1,'Technology','World')))))

In [10]:
# Now doing the same thing for Day of the Week
dataset_viz['Publish_DOW'] = np.where(dataset_viz['weekday_is_monday']==1,'Monday',np.where(dataset_viz['weekday_is_tuesday']==1,"Tuesday",np.where(dataset_viz['weekday_is_wednesday']==1,'Wednesday',np.where(dataset_viz['weekday_is_thursday']==1,'Thursday',np.where(dataset_viz['weekday_is_friday']==1,'Friday',np.where(dataset_viz['weekday_is_saturday'],'Saturday','Sunday'))))))

In [11]:
# We can go ahead and remove the columns that have been utilized
dataset_viz = dataset_viz.drop(['weekday_is_saturday','weekday_is_friday','weekday_is_sunday','weekday_is_thursday','weekday_is_wednesday','weekday_is_tuesday','weekday_is_monday','data_channel_is_lifestyle','data_channel_is_entertainment','data_channel_is_bus','data_channel_is_socmed','data_channel_is_tech','data_channel_is_world'],axis=1)

In [12]:
# Saving out this dataset for collaboration
# dataset_viz.to_csv
# We're going to use dataset_viz for visualizations and dataset for modeling
# Some of the features are dependent of particularities of the Mashable service (whose articles have been used as data source): articles often reference other articles published in the same service; and articles have meta-data, such as keywords, data channel type and total number of shares (when considering Facebook, Twitter, Google+, LinkedIn, Stumble-Upon and Pinterest). The minimum, average and maximum number of shares was determined of all Mashable links cited in the article were extracted to prepare the data. Similarly, rank of all article keyword average shares was determined, in order to get the worst, average and best keywords. For each of these keywords, the minimum, average and maximum number of shares was extracted as a feature. [Reference: Research Paper]
# Several features are extracted by performing natural language processing on the original articles. The Latent Dirichlet Allocation (LDA) algorithm was applied to all Mashable articles in order to first identify the five top relevant topics and then measure the closeness of current article to such topics. To compute the subjectivity and polarity sentiment analysis, Pattern web mining module was adopted, allowing the computation of sentiment polarity and subjectivity scores. These are such features:
# 1. Closeness to top 5 LDA topics 
# 2. Title subjectivity ratio 
# 3. Article text subjectivity score and its absolute difference to 0.5 
# 4. Title sentiment polarity 
# 5. Rate of positive and negative words 
# 6. Pos. words rate among non-neutral words 
# 7. Neg. words rate among non-neutral words 
# 8. Polarity of positive words (min./avg./max.) 
# 9. Polarity of negative words (min./avg./max.) 
# 10. Article text polarity score and its absolute difference to 0.5
# [Reference: Research Paper]

In [13]:
dataset_viz.to_csv('/Users/manishatherupalli/Desktop/ML FINAL PROJECT/Dataset/OnlineNewsPopularity_Viz.csv')